File: src/main/java/study/codingtest/leetcode/divide-conquer/LC_241.ts
const memo: Map<string, number[]> = new Map();

const diffWaysToCompute = (expression: string): number[] => {
  if (memo.has(expression)) return memo.get(expression);

  const result: number[] = [];
  for (let i = 0; i < expression.length; i++) {
    const char = expression[i];
    if (/[+\-*]/gi.test(char)) {
      const left = diffWaysToCompute(expression.substring(0, i));
      const right = diffWaysToCompute(expression.substring(i + 1));

      for (const l of left) {
        for (const r of right) {
          if (char === "+") result.push(l + r);
          else if (char === "-") result.push(l - r);
          else if (char === "*") result.push(l * r);
        }
      }
    }
  }

  if (result.length === 0) result.push(parseInt(expression));
  memo.set(expression, result);
  return result;
};

File: src/main/java/study/codingtest/leetcode/divide_n_conquer/LC_241.kt
val memo: MutableMap<String, List<Int>> = mutableMapOf()

fun diffWaysToCompute(expression: String): List<Int> {
  val result: MutableList<Int> = mutableListOf()

  if (memo.containsKey(expression)) {
    return memo[expression]!!
  }

  for (i in expression.indices) {
    val c = expression[i]
    if (c in "+-*") {
      val left = diffWaysToCompute(expression.substring(0, i))
      val right = diffWaysToCompute(expression.substring(i + 1))

      for (l in left) {
        for (r in right) {
          when (c) {
            '+' -> result.add(l + r)
            '-' -> result.add(l - r)
            '*' -> result.add(l * r)
          }
        }
      }
    }
  }

  if (result.isEmpty()) {
    result.add(expression.toInt())
  }
  memo[expression] = result
  return result
}